### PR TITLE
[ADD] Mesa Purchase Module

### PR DESCRIPTION
**Pull request by slsanc**
**[#12805 Auto populate Sales rep on Bill's based on PO [15505]](https://erp.bluestingray.com/web#id=15505&view_type=form&model=project.task&action=378&menu_id=263)**
**[#12816 Auto populate Sales rep on PO's based on Related Customer [15504]](https://erp.bluestingray.com/web#id=15504&view_type=form&model=project.task&action=378&menu_id=263)**

This commit creates the mesa_purchase module. This module adds the following features:
- When a user creates a Vendor Bill from a Purchase Order, the default Sales Representative for the Vendor Bill is now the Sales Representative of the Purchase Order.
- When a user changes the Related Customer of a Purchase Order, the Sales Representative of the Purchase Order is set to the new Related Customer’s Sales Representative.

### CODE REVIEW

**Review by traviswaelbro — COMMENTED** [review]
I haven't reviewed this yet, but a couple of early comments.

1. We want to avoid having modules which are too broad/generic and I'm worried we'll fall into that trap by having this be `mesa_purchase`, "Mesa specific purchase features". Instead, can we tailor the naming and description of these to be a bit more specific/focused, such as `mesa_purchase_default_salesperson` or similar?
2. What happens if you try to install this on a database without `x_studio_related_customer`?

File: mesa_purchase/models/purchase/order.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    force_user_id = fields.Many2one("res.users", string="Responsible Salesperson (Custom)", store=True)
+
+    @api.multi
+    def action_view_invoice(self):
+        """
+        This function returns an action that display existing vendor bills of
+        given purchase order ids.
+
+        When only one found, show the vendor bill immediately.
+        """
```
Comment (toddythebody) [inline mesa_purchase/models/purchase/order.py:17]: You still need to have your `:return` in there even if you're using core code.  Also if you're overriding, add `### Custom Code Begin ###` and `### Custom Code End ###` around your code.

File: mesa_purchase/models/purchase/order.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
+from odoo import api, fields, models
+
+
+class PurchaseOrder(models.Model):
+    _inherit = "purchase.order"
+
+    user_id = fields.Many2one(compute="_compute_user_id", inverse="_inverse_user_id")
+    force_user_id = fields.Many2one("res.users", string="Responsible Salesperson (Custom)", store=True)
```
[the remark this answers is not in the document]
Comment (slsanc) [reply mesa_purchase/models/purchase/order.py:8]: I just looked elsewhere and see that `some_field_id` and `force_some_field_id` are generally both stored. I will make it so here.

**Comment by slsanc** [comment]
> I haven't reviewed this yet, but a couple of early comments.
> 
>     1. We want to avoid having modules which are too broad/generic and I'm worried we'll fall into that trap by having this be `mesa_purchase`, "Mesa specific purchase features". Instead, can we tailor the naming and description of these to be a bit more specific/focused, such as `mesa_purchase_default_salesperson` or similar?
> 

@traviswaelbro  I was going off of the copper_inventory module and others I'd seen in our codebase. But I understand now that those are considered legacy, and we have changed practices.

>     2. What happens if you try to install this on a database without `x_studio_related_customer`?

The install fails. But we need to either reference that field here or replace the studio field with a field in the code. I started doing this, but I realized that:
1 - Replacing even one of the studio fields would mean having to replace them all, because so many of the studio fields rely on each other.
2 - the developers at Odoo are working around these studio fields, so we probably can't undertake the task of replacing the studio fields until their time has run out.

File: mesa_purchase/models/purchase/order.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    @api.multi
+    @api.depends("x_studio_related_customer")
+    def _compute_user_id(self):
+        """
+        Compute the Sales Rep to be the Related Customer's Sales Rep.
+
+        :return {NoneType}:
+        """
+        for purchase_id in self:
+            purchase_id.user_id = purchase_id.x_studio_related_customer.user_id.id
```
[the remark this answers is not in the document]
Comment (slsanc) [reply mesa_purchase/models/purchase/order.py]: I'm not quite sure what you mean. It is being declared on line 8, then used in the inverse_user_id function. Should I take out line 8?

File: mesa_purchase_default_salesperson/__manifest__.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
+# noinspection PyStatementEffect
+{
+    'name': 'Mesa Purchase Default Salesperson',
+    'category': 'Purchase',
+    'version': '12.0.0',
```
Comment (traviswaelbro) [inline mesa_purchase_default_salesperson/__manifest__.py:5]: Why isn't black running automatically?

```suggestion
    "name": "Mesa Purchase Default Salesperson",
    "category": "Purchase",
    "version": "12.0.0",
```

File: mesa_purchase_default_salesperson/models/purchase/order.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    @api.multi
+    def action_view_invoice(self):
+        """
+        This function displays existing vendor bills of given purchase order
+        ids. If only one is found, show the vendor bill immediately. If
+        create_bill is true, it will create the vendor bill.
+
+        :return: an action that displays vendor bills of
+        given purchase order ids.
+        """
```
Comment (traviswaelbro) [inline mesa_purchase_default_salesperson/models/purchase/order.py:19]: You need to write your own docstring when you're inheriting/overriding a method. It should explain whether you're inheriting or overriding and why you're doing that or what you're changing.

File: mesa_purchase_default_salesperson/models/purchase/order.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        if len(self.invoice_ids) > 1 and not create_bill:
+            result["domain"] = "[('id', 'in', " + str(self.invoice_ids.ids) + ")]"
+        else:
+            res = self.env.ref("account.invoice_supplier_form", False)
+            form_view = [(res and res.id or False, "form")]
+            if "views" in result:
+                result["views"] = form_view + [(state, view) for state, view in action["views"] if view != "form"]
+            else:
+                result["views"] = form_view
+                ### Custom Code Begin ###
```
Comment (traviswaelbro) [inline mesa_purchase_default_salesperson/models/purchase/order.py:41]: We usually use

```
### BEGIN CUSTOM CODE ###
...
### END CUSTOM CODE ###
```

File: mesa_purchase_default_salesperson/models/purchase/order.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+        :return {NoneType}:
+        """
+        for purchase_id in self:
+            purchase_id.user_id = purchase_id.x_studio_related_customer.user_id.id
+
+    @api.multi
+    def _inverse_user_id(self):
+        """
+        This allows users to manually set the LOB field in the form.
```
Comment (traviswaelbro) [inline mesa_purchase_default_salesperson/models/purchase/order.py:69]: I don't think Mesa uses LOB 😉

**Review by traviswaelbro — CHANGES_REQUESTED** [review]
Couple of minor comments. Functionally tested and looking good overall. 👍

File: mesa_purchase_default_salesperson/models/purchase/order.py (hunk 7 above)
Comment (toddythebody) [reply mesa_purchase_default_salesperson/models/purchase/order.py:69]: Maybe we should introduce it to them.  Seems to be a great concept for our other client.

File: mesa_purchase_default_salesperson/models/purchase/order.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    force_user_id = fields.Many2one("res.users", string="Responsible Salesperson (Custom)")
+
+    @api.multi
+    def action_view_invoice(self):
+        """
+        This function displays existing vendor bills of given purchase order
+        ids. If only one is found, show the vendor bill immediately. If
+        create_bill is true, it will create the vendor bill.
+
+        :return: an action that displays vendor bills of
```
Comment (toddythebody) [inline mesa_purchase_default_salesperson/models/purchase/order.py:17]: What are the types?  What are the possible actions?  What causes each action?

**Review by toddythebody — CHANGES_REQUESTED** [review]

File: mesa_purchase_default_salesperson/models/purchase/order.py (hunk 7 above)
Comment (slsanc) [reply mesa_purchase_default_salesperson/models/purchase/order.py:69]: @traviswaelbro @toddythebody 

This is terribly embarrassing. :flushed: 
To be completely honest, I had copied and pasted the inverse function from a copper module. I need to be very careful about doing this.

**Review by toddythebody — APPROVED** [review]

**Comment by slsanc** [comment]
@toddythebody that I made the force_user_id field stored, based on Aaron's advice regarding force fields [here](https://github.com/gobluestingray/mlgmain-odoo/pull/4#discussion_r595586547)

**Review by traviswaelbro — CHANGES_REQUESTED** [review]

**Comment by slsanc** [comment]
This has been moved to:
https://github.com/gobluestingray/mlgmain-odoo/pull/5

**Pull request closed without merging**